src/qmapshack/gis/search/CSearchLineEdit.cpp

```diff
@@ -0,0 +1,202 @@
+    connect(actionHelp, &QAction::triggered, this, &CSearchLineEdit::slotSearchHelp);
+    connect(actionClearFilter, &QAction::triggered, this, &CSearchLineEdit::slotClearFilter);
+    connect(this, &CSearchLineEdit::textChanged, this, &CSearchLineEdit::slotCreateSearch);
+}
+
+CSearchLineEdit::CSearchLineEdit(QWidget *parent, IGisProject* project, CSearch* search) : CSearchLineEdit(parent)
+{
+    connectedProject=project;
+    if(search != nullptr)
+    {
+        setText(search->getSearchText());
+    }
+}
+void CSearchLineEdit::slotSetupSearch()
+{
+    QMenu * menu = new QMenu(this);
+    menu->addSection(tr("Apply filter to"));
+    menu->addAction(actionNameOnly);
+    menu->addAction(actionCompleteText);
+    menu->addSection(tr("Case sensitivity"));
+    menu->addAction(actionCaseSensitive);
+
+    QActionGroup* actionGroup = new QActionGroup(menu);
+    actionGroup->addAction(actionNameOnly);
+    actionGroup->addAction(actionCompleteText);
+
+    switch(CSearch::getSearchMode())
+    {
+    case CSearch::eSearchModeName:
+        actionNameOnly->setChecked(true);
+        break;
+
+    case CSearch::eSearchModeText:
+        actionCompleteText->setChecked(true);
+        break;
+    }
+
+    if(CSearch::getCaseSensitivity() == Qt::CaseSensitive)
+    {
+        actionCaseSensitive->setChecked(true);
+    }
+    else
+    {
+        actionCaseSensitive->setChecked(false);
+    }
+    menu->move(parentWidget()->mapToGlobal(geometry().topLeft()));
+    menu->exec();
+}
+
+void CSearchLineEdit::slotSearchHelp()
+{
+    if(explanationDlg == nullptr)
+    {
+        explanationDlg = new CSearchExplanationDialog(this);
+    }
+
+    explanationDlg->setVisible(true);
+}
+
+void CSearchLineEdit::slotClearFilter()
+{
+    setText("");
+    if(connectedProject != nullptr)
+    {
+        connectedProject->setProjectFilter(CSearch(""));
+        CCanvas::triggerCompleteUpdate(CCanvas::eRedrawGis);
+    }
+    else
+    {
+        emit searchCleared(connectedProject);
+    }
+}
+
+
+void CSearchLineEdit::slotSearchNameOnly(bool yes)
+{
+    if(yes)
+    {
+        CSearch::setSearchMode(CSearch::eSearchModeName);
+        //Trigger reprocessing of search
+        slotCreateSearch(text());
+    }
+}
+
+void CSearchLineEdit::slotSearchCompleteText(bool yes)
+{
+    if(yes)
+    {
+        CSearch::setSearchMode(CSearch::eSearchModeText);
+        //Trigger reprocessing of search
+        slotCreateSearch(text());
+    }
+}
+
+void CSearchLineEdit::slotCaseSensitive(bool yes)
+{
+    if(yes)
+    {
+        CSearch::setCaseSensitivity(Qt::CaseSensitive);
+    }
+    else
+    {
+        CSearch::setCaseSensitivity(Qt::CaseInsensitive);
+    }
+    //Trigger reprocessing of search
+    slotCreateSearch(text());
+}
+
+void CSearchLineEdit::slotCreateSearch(const QString& str)
+{
+    actionClearFilter->setIcon(str.isEmpty() ? QIcon("://icons/32x32/Filter.png") : QIcon("://icons/32x32/Cancel.png"));
```

Comment:
I think we have to set 
```
        CCanvasCursorLock cursorLock(Qt::WaitCursor, __func__);
        QMutexLocker lock(&IGisItem::mutexItems);
```

somewhere in this method. But I am not sure where to place it. It shouldn't interfere with

```
void CGisWorkspace::slotSearch(const CSearch& currentSearch)
```

Reply:
A bit more background information:

The draw thread is iterating over the workspace tree widget to draw all items. If that happens it's a bad idea to alter the tree widget or to access items. That's why each operation on the workspace tree widget should be blocked by the `IGisItem` mutex. 

Showing a wait cursor is just good practice.

Reply:
What about placing it into `IGisProject::applyFilters()`, where the actual manipulation happens?